### PR TITLE
use http over stdio to update config

### PR DESCRIPTION
**Pull request by Hexilee**
Signed-off-by: hexilee <i@hexilee.me>

### Features

- use HTTP over stdio to update config
As mentioned in [chaos-mesh#1706](https://github.com/chaos-mesh/chaos-mesh/pull/1706#discussion_r621766031), we need the update result from stdio.

- ensure actions order
As mentioned in [chaos-mesh#1706](https://github.com/chaos-mesh/chaos-mesh/pull/1706#discussion_r623610580), we should ensure the order of the action applied.

- rename append action to patch action; support json merge patch as [rfc7396](https://tools.ietf.org/html/rfc7396)

- support wildcard URI path as rules of [wildcard matches](https://www.wikiwand.com/en/Matching_wildcards).

#### Build
```bash
> cargo build --release
```

#### Run and apply config
```bash
> sudo target/release/tproxy -i
PUT / HTTP/1.1
Content-Length: 129

{"proxy_ports": [30086], "rules": [{"target": "Request", "selector": {"path": "*", "method": "GET"},"actions": {"abort": true}}]}
```

get the response:

```bash
HTTP/1.1 200 OK
content-length: 0
date: Mon, 03 May 2021 11:21:31 GMT
```

```
PUT / HTTP/1.1
Content-Length: 129

{"proxy_ports": [30086], "rules": [{"target": "Request", "selector": {"path": "*", "method": "GET"},"actions": {"abort": true}}]}
HTTP/1.1 200 OK
content-length: 0
date: Mon, 03 May 2021 11:21:31 GMT

PUT / HTTP/1.1
Content-Length: 124

{"proxy_ports": [], "rules": [{"target": "Request", "selector": {"path": "*", "method": "GET"},"actions": {"abort": true}}]}
HTTP/1.1 200 OK
content-length: 0
date: Mon, 03 May 2021 11:22:13 GMT
```

#### Exit
Ctrl-C

### CODE REVIEW

**Comment by Andrewmatilde** [comment]
Now if user just use to command line like ```sudo ./target/release/tproxy``` will not print any information .
So I suggest to add a flag to keep STDIN open, just like docker run.
```
-i, --interactive Keep STDIN open even if not attached
```
And it is necessary to add more information to Readme.

File: src/config_server.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    type Future = BoxedSendFuture<Self::Response, Self::Error>;
+
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    #[inline]
+    fn call(&mut self, _: &StdStream) -> Self::Future {
+        let service = self.clone();
+        Box::pin(async move { Ok(service) })
```
Comment (Andrewmatilde) [inline src/config_server.rs:125]: I am confused about how std stream works here , since there is not handler just like Request<Body> or tcpstream.

Comment (Hexilee) [reply src/config_server.rs:125]: The hyper server will read/write the stream returned by `Accept` automatically. Here is just an immutable reference to record some session info.

**Comment by Hexilee** [comment]
> 
> 
> Now if user just use to command line like `sudo ./target/release/tproxy` will not print any information .
> So I suggest to add a flag to keep STDIN open, just like docker run.
> 
> ```
> -i, --interactive Keep STDIN open even if not attached
> ```
> 
> And it is necessary to add more information to Readme.

Sounds reasonable.

File: src/config_server.rs (hunk 1 above)
Comment (Andrewmatilde) [reply src/config_server.rs:125]: Thx

**Comment by Hexilee** [comment]
> 
> 
> Now if user just use to command line like `sudo ./target/release/tproxy` will not print any information .
> So I suggest to add a flag to keep STDIN open, just like docker run.
> 
> ```
> -i, --interactive Keep STDIN open even if not attached
> ```
> 
> And it is necessary to add more information to Readme.

Implemented. PTAL @Andrewmatilde @YangKeao

File: src/main.rs [hunk 2]
```diff
@@ -1,66 +1,43 @@
...
-                }
-                if let Err(err) = reload(&mut server, &buf).await {
-                    error!("error in reloading http server: {}", err);
-                    break;
-                }
-            }
-        };
-    }
-
+    signal_handler.wait().await?;
```
Comment (YangKeao) [inline src/main.rs:40]: Why don't use HTTP to stop?

Comment (Hexilee) [reply src/main.rs:40]: Because stopping by signals is more convenient, routes will be reset if you stop it by SIGINT accidentally.

**Review by YangKeao — APPROVED** [review]
LGTM

**Comment by STRRL** [comment]
Oops, it seems the magic word "resolve" will do something unexpected. 🌚

image

**Comment by Hexilee** [comment]
> 
> 
> Oops, it seems the magic word "resolve" will do something unexpected. 🌚
> 
> image

???, changed 😂

File: src/cmd.rs [hunk 3]
```diff
@@ -5,29 +5,56 @@ use anyhow::{anyhow, Result};
...
+    pub input: Option<PathBuf>,
 }
 
-pub async fn get_config() -> Result<Config> {
-    let opt: Opt = Opt::from_args();
-    get_config_from_opt(opt.input).await
+impl Opt {
+    pub fn get_level_filter(&self) -> LevelFilter {
+        match self.verbose {
+            0 => LevelFilter::ERROR,
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply src/cmd.rs:34]: I think they are unnecessary, as we do not use `WARN`, and I think errors should always be printed out.

**Review by Andrewmatilde — APPROVED** [review]

**Comment by Andrewmatilde** [comment]
LGTM

**Pull request merged**